### PR TITLE
Refactored device_disasm_interface::disassemble() and dasm_override_delegate to take string buffers as std::string

### PR DESCRIPTION
**Pull request by npwoods**
Also consolidated dasm_override_delegates for CoCo and Dragon Beta

Feel free to defer this until after the release

### CODE REVIEW

**Comment by galibert** [comment]
Urgh, that's very much done halfways.  Do you plan to change the
disasm_disassemble signature in a followup pull request?

  OG.

**Comment by cuavas** [comment]
Also you shouldn't be using a `std::string` as a text sink as it's not designed for this, and not particularly efficient. You should use `std::ostream` or something like it. This won't be merged in its current form.

**Comment by npwoods** [comment]
Updated the string sinks to be std::ostream

Updating the actual disasm_disassemble() calls to use std::ostream would be straight forward; that can easily be done as a follow up provided that this patch passes muster.

**Comment by npwoods** [comment]
are there any outstanding concerns about this change?

**Comment by cuavas** [comment]
Well, it's a very big changeset that requires pulling locally and reading over slowly in a git client, and I haven't had a chance to do that yet.  But if it still contains lots of static `char []` buffers, then it won't be merged in its current state, because if it is, it will inevitably rot and not get cleaned up properly, because that's what happens when things are half-done.

**Comment by npwoods** [comment]
I believe that there are still a lot of 'char[]' arrays in the disassemblers, but not many 'static char[]' array anymore.

Also a note - in a separate branch, I've prepared changes that merge the 6809/6309/Konami disassemblers into a single module.  I'm keeping that separate from this branch because I don't want to increase the footprint of this changeset, and on top of that I want to keep the Konami changes (which actually alter the behavior of the disassembler with some bug fixes) distinct.

**Comment by Cowering** [comment]
Yay, that will get rid of several more ODR errors that LTO throws.

**Comment by cuavas** [comment]
Could you rebase and squash this to make it a bit more manageable?

File: src/devices/cpu/powerpc/ppc_dasm.cpp [hunk 1]
```diff
@@ -1169,18 +1169,19 @@ offs_t ppc_dasm_one(std::ostream &stream, uint32_t pc, uint32_t op)
 	return 4 | flags;
 }
 
-offs_t ppc_dasm_one(char *buffer, uint32_t pc, uint32_t op)
+
+offs_t ppc_dasm_one(std::string &string, uint32_t pc, uint32_t op)
```
Comment (cuavas) [inline src/devices/cpu/powerpc/ppc_dasm.cpp]: Can you eliminate this extra trampoline altogether?  It doesn't seem to serve much of a purpose.

File: src/devices/cpu/powerpc/ppcdrc.cpp [hunk 2]
```diff
@@ -3926,20 +3926,20 @@ void ppc_device::log_opcode_desc(drcuml_state *drcuml, const opcode_desc *descli
 	/* output each descriptor */
 	for ( ; desclist != nullptr; desclist = desclist->next())
 	{
-		char buffer[100];
+		std::string buffer;
```
Comment (cuavas) [inline src/devices/cpu/powerpc/ppcdrc.cpp]: Could you turn this into some kind of stream object and lift it out of the loop?  Then allocations can be saved.

File: src/devices/cpu/rsp/rsp_dasm.cpp [hunk 3]
```diff
@@ -341,12 +341,11 @@ offs_t rsp_dasm_one(std::ostream &stream, offs_t pc, uint32_t op)
 	return 4 | flags | DASMFLAG_SUPPORTED;
 }
 
-offs_t rsp_dasm_one(char *buffer, offs_t pc, uint32_t op)
```
Comment (cuavas) [inline src/devices/cpu/rsp/rsp_dasm.cpp]: As with PPC, this trampoline isn't really very useful - the code without it isn't any more ugly.

File: src/devices/cpu/sharc/sharcops.hxx [hunk 4]
```diff
@@ -2722,8 +2722,8 @@ void adsp21062_device::sharcop_idle()
 void adsp21062_device::sharcop_unimplemented()
 {
 	extern CPU_DISASSEMBLE(sharc);
-	char dasm[1000];
+	std::stringstream dasm;
```
Comment (cuavas) [inline src/devices/cpu/sharc/sharcops.hxx]: Does this stream need to be input-capable?

**Comment by cuavas** [comment]
Please make the vgmplay disassembler use streams properly, not cheat and do it backwards.

File: src/devices/cpu/powerpc/ppcdrc.cpp [hunk 5]
```diff
@@ -3919,27 +3919,31 @@ void ppc_device::log_register_list(drcuml_state *drcuml, const char *string, con
...
 			else
 				ppc_dasm_one(buffer, desclist->pc, desclist->opptr.l[0]);
 		}
 		else
-			buffer = "???";
+			buffer << "???";
 
-		drcuml->log_printf("%08X [%08X] t:%08X f:%s: %-30s", desclist->pc, desclist->physpc, desclist->targetpc, log_desc_flags_to_string(desclist->flags), buffer.c_str());
+		buffer.put('\0');
+		drcuml->log_printf("%08X [%08X] t:%08X f:%s: %-30s", desclist->pc, desclist->physpc, desclist->targetpc, log_desc_flags_to_string(desclist->flags), buffer.vec());
```
Comment (cuavas) [inline src/devices/cpu/powerpc/ppcdrc.cpp]: I think it needs to be &buffer.vec()[0] or you'll get a problem trying to pass an object with varargs.

File: src/emu/debug/debugcmd.cpp [hunk 6]
```diff
@@ -2429,6 +2429,8 @@ void debugger_commands::execute_dasm(int ref, int params, const char *param[])
 			}
 
 			/* disassemble the result */
+			disasm.clear();
+			disasm.seekp(0);
```
Comment (cuavas) [inline src/emu/debug/debugcmd.cpp]: this needs to be moved outside the `if` statement - the string was re-constructed each time through the loop, but now it won't be.

**Pull request merged**